File: typescript/lld/vending-machine/State.ts
import type { VendingMachine } from './VendingMachine';

// The contract every state must honour. Each action defaults to
// "not allowed here", so a concrete state overrides only its LEGAL actions.
export abstract class State {
    insertCoin(m: VendingMachine, cents: number): void  { this.reject('insert coin'); }
    selectProduct(m: VendingMachine, id: string): void  { this.reject('select product'); }
    dispense(m: VendingMachine): void                   { this.reject('dispense'); }
    refund(m: VendingMachine): void                     { this.reject('refund'); }

    abstract name(): string;

    protected reject(action: string): void {
        console.log(`  x cannot ${action} while ${this.name()}`);
    }
}

File: java/lld/vending-machine/State.java
// The contract every state must honour. Each action defaults to
// "not allowed here", so a concrete state overrides only its LEGAL actions.
public abstract class State {
    public void insertCoin(VendingMachine m, int cents)    { reject("insert coin"); }
    public void selectProduct(VendingMachine m, String id) { reject("select product"); }
    public void dispense(VendingMachine m)                 { reject("dispense"); }
    public void refund(VendingMachine m)                   { reject("refund"); }

    public abstract String name();

    protected void reject(String action) {
        System.out.println("  x cannot " + action + " while " + name());
    }
}